modules/caddyhttp/reverseproxy/httptransport.go

```diff
@@ -510,6 +527,28 @@ type KeepAlive struct {
 	IdleConnTimeout caddy.Duration `json:"idle_timeout,omitempty"`
 }
 
+// tcpRWTimeoutConn enforces read/write timeouts for a TCP connection.
+type tcpRWTimeoutConn struct {
+	*net.TCPConn
+	readTimeout, writeTimeout time.Duration
+}
+
+func (c *tcpRWTimeoutConn) Read(b []byte) (int, error) {
+	err := c.TCPConn.SetDeadline(time.Now().Add(c.readTimeout))
+	if err != nil {
+		return 0, err
+	}
```

Comment:
One way to address the problem mentioned above is to set the deadline only when timeout is non-zero. In addition, I guess `SetReadDeadline` is enough here? For example:

```go
if d := c.readTimeout; d > 0 {
    err := c.TCPConn.SetReadDeadline(time.Now().Add(d))
    if err != nil {
        return 0, err
    }
}
```

By the way, I even found that [the error result is unhandled in readRequest of net/http](https://github.com/golang/go/blob/c5da4fb7ac5cb7434b41fc9a1df3bee66c7f1a4d/src/net/http/server.go#L981-L986).

Reply:
Indeed. And that may be because we can't do anything useful if there is an error. :man_shrugging: Especially not in the general case of standard library code :sweat_smile: 

We can perhaps at least log if setting the deadline fails.